### ISSUE TITLE
Generic corrupt file handling

### ISSUE THREAD

**Issue by Joshfindit** [kaitai-io/kaitai_struct#843]
Big respect for the goals of project.
I feel like Kaitai may even be the end-game documentation method for binary files.

One thing I can say across decades of work with files is that some type of corruption is inevitable.

**Does the parser have any sort of handling for this?**

- Does it fail gracefully during file read?
- Is everything up to the corruption kept so that the user can process it?
- Does it continue reading after the corruption in an attempt to recover anything around it?
- Are the answers to these questions language-agnostic?

**Comment by KOLANICH** [comment]
No × 4.

No builtin handling if errorrs. But there is a possibility to disable automatic parsing and do it oneself. Though doing that you lose the main benefit KS gives you that you don't have to deal with parsing yourself.

There is no syntax to specify recovery.

But you can do the following trick.

1. generate the source
2. edit it to incorporate errors handling
3. generate a patch file.

Then it may be possible to apply the patch file to the generated code.

**Comment by KOLANICH** [comment]
@generalmimon, don't you want to activate discussions for the repo

**Comment by generalmimon** [comment]
@Joshfindit 

> Is everything up to the corruption kept so that the user can process it?

@KOLANICH 

> No × 4.
> 
> **No builtin handling if errorrs**. But there is a possibility to disable automatic parsing and do it oneself. Though doing that you lose the main benefit KS gives you that you don't have to deal with parsing yourself.

Not quite true. If you compile the KSY spec with `--no-auto-read`, you can just wrap the `_read()` call into a `try..catch` block and even if you catch an error, the data parsed before the error will be available in the parser object. See https://github.com/kaitai-io/kaitai_struct/issues/761#issuecomment-648340184 for more details.

**Comment by KOLANICH** [comment]
>Not quite true. If you compile the KSY spec with --no-auto-read, you can just wrap the _read() call into a try..catch block and even if you catch an error, the data parsed before the error will be available in the parser object.

Yes,  not "everything", the data is not discarded. But it is not recovery in the sense that we cannot resynchronise and continue reading the data. 

To implement user-assisted recovery we need to generate a finite state machine and allow a user to set its state, and it should be implemented the way to make it easy to use.

1. introduce a syntax for checkpoints within `seq` blocks. Each checkpoint has a name.
2. KSC cuts the parsing function into multiple functions. Each function in the end sets the state pointing to the next function.
3. `_parse` method just initializes with the first function and then calls another method that just iterates until the state pointer is null
4. if an error occurs, then iteration stops, and a user can get the state at checkpoint granularity. than the user recovers manually with his custom code, and sets the state to the next checkpoint. then he calls the iterating method.
5 the cutting  should occur only if there are checkpoints.

But I have a feeling like it would make KS Turing-complete.

**Comment by generalmimon** [comment]
> But it is not recovery in the sense that we cannot resynchronise and continue reading the data.

Sure, it's not.

> 1. introduce a syntax for checkpoints within `seq` blocks. Each checkpoint has a name.

When you speak of _checkpoints_ and _states_, do you mean something like https://github.com/kaitai-io/kaitai_struct/issues/449#issuecomment-394894567?

**Comment by KOLANICH** [comment]
Yes, something like that, but different, and let's continue there then.
